context_test.go

```diff
@@ -282,3 +284,46 @@ func createTestMutator(ctx TopDownMutatorContext) {
 		Name: "D",
 	})
 }
+
+// test that WalkBlueprintsFiles calls asyncVisitor in the right order
+func TestWalkFileOrder(t *testing.T) {
+	// setup mock context
+	ctx := newContext()
+	mockFiles := map[string][]byte{
+		"Blueprints": []byte(`
+			sample_module {
+			    name: "a",
+			}
+		`),
+		"dir1/Blueprints": []byte(`
+			sample_module {
+			    name: "b",
+			}
+		`),
+		"dir1/dir2/Blueprints": []byte(`
+			sample_module {
+			    name: "c",
+			}
+		`),
+	}
+	ctx.MockFileSystem(mockFiles)
+
+	// prepare to monitor the visit order
+	visitOrder := []string{}
+	visitLock := sync.Mutex{}
+	visitor := func(file *parser.File) {
+		visitLock.Lock()
```

Comment:
Might be worth adding a sleep or something in the parent to make sure the don't just happen to finish in the right order when running them asynchronously.

Reply:
sounds good

Reply:
I don't think a constant sleep will work, you need the goroutine that is started first to sleep longer than the one that is started second to ensure that the one that is started second can never run first.

Reply:
It turns out that the previous race condition inside WalkBlueprintsFiles applied to the order that visit was called, but that visit itself was called synchronously. As a result, adding some sleeps doesn't cause this test to fail (in practice) with the old code either.

Although I've now updated the sleep anyway.